### ISSUE TITLE
Test failure: org.g_node.nix.TestUtil.testSIUnitSplit(TestUtil.java:84)

### ISSUE THREAD

**Issue by gicmo** [G-Node/nix-java#66]
```bash
Running org.g_node.nix.TestUtil
[nix::util::convertToSeconds] Warning: given unit is not supported!
[nix::util::convertToKelvin] Warning: given unit is not supported
Tests run: 10, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.096 sec <<< FAILURE! - in org.g_node.nix.TestUtil
testSIUnitSplit(org.g_node.nix.TestUtil)  Time elapsed: 0.016 sec  <<< FAILURE!
java.lang.AssertionError
	at org.g_node.nix.TestUtil.testSIUnitSplit(TestUtil.java:84)
```

**Comment by gicmo** [comment]
Ok, here is my analysis, not sure it is correct, would love if someone else gives it some eyes:

The generated code for `splitUnit` is this. The whole code is in this [gist](https://gist.github.com/gicmo/1a6d47c292ee1e2c4917):

```C++
JNIEXPORT void JNICALL Java_org_g_1node_nix_util_Util_splitUnit(JNIEnv* env, jclass cls, jstring arg0, jobject arg1, jobject arg2, jobject arg3) {
    const char* ptr0 = arg0 == NULL ? NULL : env->GetStringUTFChars(arg0, NULL);
    jint size0 = 0;
    void* owner0 = (void*)ptr0;
    StringAdapter adapter0(ptr0, size0, owner0);
    signed char* ptr1 = arg1 == NULL ? NULL : (signed char*)jlong_to_ptr(env->GetLongField(arg1, JavaCPP_addressFID));
    jint size1 = arg1 == NULL ? 0 : env->GetIntField(arg1, JavaCPP_limitFID);
    void* owner1 = JavaCPP_getPointerOwner(env, arg1);
    jint position1 = arg1 == NULL ? 0 : env->GetIntField(arg1, JavaCPP_positionFID);
    ptr1 += position1;
    size1 -= position1;
    StringAdapter adapter1(ptr1, size1, owner1);
    signed char* ptr2 = arg2 == NULL ? NULL : (signed char*)jlong_to_ptr(env->GetLongField(arg2, JavaCPP_addressFID));
    jint size2 = arg2 == NULL ? 0 : env->GetIntField(arg2, JavaCPP_limitFID);
    void* owner2 = JavaCPP_getPointerOwner(env, arg2);
    jint position2 = arg2 == NULL ? 0 : env->GetIntField(arg2, JavaCPP_positionFID);
    ptr2 += position2;
    size2 -= position2;
    StringAdapter adapter2(ptr2, size2, owner2);
    signed char* ptr3 = arg3 == NULL ? NULL : (signed char*)jlong_to_ptr(env->GetLongField(arg3, JavaCPP_addressFID));
    jint size3 = arg3 == NULL ? 0 : env->GetIntField(arg3, JavaCPP_limitFID);
    void* owner3 = JavaCPP_getPointerOwner(env, arg3);
    jint position3 = arg3 == NULL ? 0 : env->GetIntField(arg3, JavaCPP_positionFID);
    ptr3 += position3;
    size3 -= position3;
    StringAdapter adapter3(ptr3, size3, owner3);
    jthrowable exc = NULL;
    try {
        nix::util::splitUnit((std::string&)adapter0, (std::string&)adapter1, (std::string&)adapter2, (std::string&)adapter3);
    } catch (...) {
        exc = JavaCPP_handleException(env, 17);
    }

    if (arg0 != NULL) env->ReleaseStringUTFChars(arg0, ptr0);
    signed char* rptr1 = adapter1;
    jint rsize1 = (jint)adapter1.size;
    void* rowner1 = adapter1.owner;
    if (rptr1 != ptr1) {
        JavaCPP_initPointer(env, arg1, rptr1, rsize1, rowner1, &StringAdapter::deallocate);
    } else {
        env->SetIntField(arg1, JavaCPP_limitFID, rsize1 + position1);
    }
    signed char* rptr2 = adapter2;
    jint rsize2 = (jint)adapter2.size;
    void* rowner2 = adapter2.owner;
    if (rptr2 != ptr2) {
        JavaCPP_initPointer(env, arg2, rptr2, rsize2, rowner2, &StringAdapter::deallocate);
    } else {
        env->SetIntField(arg2, JavaCPP_limitFID, rsize2 + position2);
    }
    signed char* rptr3 = adapter3;
    jint rsize3 = (jint)adapter3.size;
    void* rowner3 = adapter3.owner;
    if (rptr3 != ptr3) {
        JavaCPP_initPointer(env, arg3, rptr3, rsize3, rowner3, &StringAdapter::deallocate);
    } else {
        env->SetIntField(arg3, JavaCPP_limitFID, rsize3 + position3);
    }
    if (exc != NULL) {
        env->Throw(exc);
    }
}
```
And calling `splitUnit("V")` with `BytePointer(1)`, i.e. the result bytePointer can hold one byte, leads to the resulting `BytePointer` as follows:

```
org.bytedeco.javacpp.BytePointer[address=0x7f9d93850050,position=0,limit=0,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x7f9d93850050,deallocatorAddress=0x1283d8cd0]]
org.bytedeco.javacpp.BytePointer[address=0x7f9d9384e4f0,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x7f9d9384e4f0,deallocatorAddress=0x1283d8cd0]]
org.bytedeco.javacpp.BytePointer[address=0x7f9d9386fba0,position=0,limit=0,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x7f9d9386fba0,deallocatorAddress=0x1283d8cd0]]
```

But `BytePointer.getString()` assumes "a null-terminated string if {@code limit <= position}", i.e. for our `BytePointers` `bp1` and `bp3`. `nix::splitUnit` didn't did assign `""` to the strings but that did not change the size in javacpp's `StringAdapter` (because `std::string("").size() == 1` which is the size of the `BytePointer`. Therefore we thing we have a null-terminated string, but we actually have not initialised the `BytePointer` with anything (as far I can see `Java_org_bytedeco_javacpp_BytePointer_allocateArray` does not initialise the allocated char array).

If we replace `BytePointer bp1 = new BytePointer(1)` with `BytePointer bp1 = new BytePointer((Pointer) null)` we get:
```
org.bytedeco.javacpp.BytePointer[address=0x0,position=0,limit=0,capacity=0,deallocator=null]

org.bytedeco.javacpp.BytePointer[address=0x7f9cc3974370,position=0,limit=1,capacity=1,deallocator=org.bytedeco.javacpp.Pointer$NativeDeallocator[ownerAddress=0x7f9cc3974370,deallocatorAddress=0x123022480]]

org.bytedeco.javacpp.BytePointer[address=0x0,position=0,limit=0,capacity=0,deallocator=null]
```
NB:  `address=0x0` for `bp1` and `bp3`. If we then check for `isNull()` the test failures go away.
They also go away if we initialise the `BytePointer` with `""`, i.e. `BytePointer bp1 = new BytePointer("")`.
Not allocating memory that we don't need seems easier therefore I think I will go with that to fix this issue (hopefully).

**Comment by jgrewe** [comment]
this really looks plausible but what really bugs me is that it occurred after rev 1.1 the test itself is really old... 
Was there a java version change in the meantime?

**Comment by gicmo** [comment]
@jgrewe I am not entirely sure but @stoewer reported that tests have been randomly failing before (hence the disabled checks on appveyor, c.f. issue #67). I am not sure this was one of the randomly failing tests but I would be surprised, because if the analysis above is correct, we might have read uninitialised memory. I am not sure I understand the "test itself is really old" part.

**Comment by jgrewe** [comment]
well, the test is in there right from the start of the project and ist not a "new" addition to the unit tests. 

Do you really think it ever worked? Limit and capacity of the BytePointer object are clearly zero.

**Comment by gicmo** [comment]
@jgrewe before the change, they were initialised with `BytePointer(1)` leading to `capacity=1` (see above , the first three `BytePointer.toString()` rows). This results in a [`StringAdapter`](https://gist.github.com/gicmo/1a6d47c292ee1e2c4917#file-jniutil-cpp-L280) with `size==1` (I assume). 

For the failing bits of this test, the `StringAdapter::str2` member will be given to `nix::util::splitUnit` via `StringAdapter::str` (a `string&` pointing to `StringAdapter::str2`). After the call to `nix::util::splitUnit` the out strings will either have `size() == 0` for `""` or `size() == 1` for `"V"`. 
After the call, when `Java_org_g_1node_nix_util_Util_splitUnit` [converts](https://gist.github.com/gicmo/1a6d47c292ee1e2c4917#file-jniutil-cpp-L1518) the `StringAdapter` to the `signed char*` the [`str.size() > size`](https://gist.github.com/gicmo/1a6d47c292ee1e2c4917#file-jniutil-cpp-L300) check will be false, this leads to the same pointer being returned (pointing to `ByteArray`s char buffer). This in turn results in the check [`rptr1 != ptr1`](https://gist.github.com/gicmo/1a6d47c292ee1e2c4917#file-jniutil-cpp-L1521) being false; hence we only setting the `BytePointer.limit` (via [`SetIntField`](https://gist.github.com/gicmo/1a6d47c292ee1e2c4917#file-jniutil-cpp-L1524)). But we have never actually touched, or actively initialised, the memory (of size 1) that `BytePointer` is pointing to (see the allocation of it [here](https://gist.github.com/gicmo/1a6d47c292ee1e2c4917#file-jniutil-cpp-L680)).
If it happened to be zero that all is fine, if not, then the test would fail. Does that make sense?

**Comment by gicmo** [comment]
Fixed with PR #63
